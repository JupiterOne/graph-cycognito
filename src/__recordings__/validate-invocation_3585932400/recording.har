{
  "log": {
    "_recordingName": "validate-invocation",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "dee32538e28d432f581f1db5e590c99d",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 2,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "2"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "api.platform.cycognito.com"
            }
          ],
          "headersSize": 344,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/json",
            "params": [],
            "text": "[]"
          },
          "queryString": [
            {
              "name": "count",
              "value": "0"
            },
            {
              "name": "offset",
              "value": "0"
            }
          ],
          "url": "https://api.platform.cycognito.com/v1/issues?count=0&offset=0"
        },
        "response": {
          "bodySize": 2,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 2,
            "text": "[]"
          },
          "cookies": [
            {
              "expires": "2022-06-08T06:03:45.000Z",
              "httpOnly": true,
              "name": "GCLB",
              "path": "/",
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 08 Jun 2022 06:02:57 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "content-length",
              "value": "2"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "via",
              "value": "1.1 google"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "cf-cache-status",
              "value": "DYNAMIC"
            },
            {
              "name": "expect-ct",
              "value": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\""
            },
            {
              "name": "server",
              "value": "cloudflare"
            },
            {
              "name": "cf-ray",
              "value": "717f6bdd1de02e9e-SIN"
            }
          ],
          "headersSize": 408,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-06-08T06:02:54.247Z",
        "time": 2326,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 2326
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
