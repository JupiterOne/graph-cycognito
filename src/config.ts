import {
  IntegrationExecutionContext,
  IntegrationValidationError,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  accessToken: {
    type: 'string',
    mask: true,
  },
};

export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The provider Access Token
   */
  accessToken: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;

  if (!config.accessToken) {
    throw new IntegrationValidationError(
      'Config requires all of {accessToken}, you can request one from your provider.',
    );
  }

  const apiClient = createAPIClient(config);
  await apiClient.verifyAuthentication();
}
