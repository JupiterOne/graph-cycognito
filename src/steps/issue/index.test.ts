import { executeStepWithDependencies } from '@jupiterone/integration-sdk-testing';
import { buildStepTestConfigForStep } from '../../../test/config';
import { Recording, setupProjectRecording } from '../../../test/recording';
import {
  Entities,
  IntegrationSteps,
  MappedRelationships,
  Relationships,
} from '../constants';

// See test/README.md for details
let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test('fetch-issues', async () => {
  recording = setupProjectRecording({
    directory: __dirname,
    name: 'fetch-issues',
  });

  const stepConfig = buildStepTestConfigForStep(IntegrationSteps.ISSUES);

  const { collectedEntities, collectedRelationships, encounteredTypes } =
    await executeStepWithDependencies(stepConfig);

  expect(encounteredTypes).toMatchSnapshot();

  expect(
    collectedEntities.some((e) => e._type === Entities.ISSUE._type),
  ).toBeTruthy();

  expect(
    collectedRelationships.some(
      (e) => e._type === Relationships.ACCOUNT_HAS_ISSUE._type,
    ),
  ).toBeTruthy();

  expect(
    collectedRelationships.some(
      (e) => e._type === MappedRelationships.ISSUE_IS_CVE._type,
    ),
  ).toBeTruthy();

  expect(
    collectedRelationships.some(
      (e) => e._type === MappedRelationships.ISSUE_IS_CYC._type,
    ),
  ).toBeTruthy();
});

test('build-asset-has-issue-relationship', async () => {
  recording = setupProjectRecording({
    directory: __dirname,
    name: 'build-asset-has-issue-relationship',
  });

  const stepConfig = buildStepTestConfigForStep(
    IntegrationSteps.BUILD_ASSET_ISSUE_RELATIONSHIPS,
  );
  const stepResult = await executeStepWithDependencies(stepConfig);
  expect(stepResult).toMatchStepMetadata(stepConfig);
});
