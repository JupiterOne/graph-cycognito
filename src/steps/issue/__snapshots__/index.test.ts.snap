// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch-issues 1`] = `
Array [
  "cycognito_account",
  "cycognito_issue",
  "cycognito_account_has_issue",
  "cycognito_issue_is_cve",
  "cycognito_issue_is_cyc",
]
`;
